src/gt_pro.cpp

```diff
@@ -292,103 +270,390 @@ bool kmer_lookup_work(LmerRange *lmer_index, uint64_t *mmer_bloom, uint32_t *kme
+    const ssize_t bytes_read = read(fd, window + bytes_leftover, segment_size - bytes_leftover);
+    if (bytes_read == (ssize_t)-1) {
+      cerr << chrono_time() << ":  "
+           << "unknown fatal error when reading " << in_path << endl;
+      exit(EXIT_FAILURE);
+    }
+
+    const auto bytes_in_window = bytes_leftover + bytes_read;
+    if (bytes_in_window == 0) {
+      // we are done for this file
+      close(fd);
+      results[channel]->finished_reading = true;
+      if (channel > last_result_idx_done_with_input) {
+        last_result_idx_done_with_input = channel;
+      }
+      ++channel;
+      if (channel == n_inputs) {
+        finished_reading_all_input = true;
+      } else {
+        in_path = input_paths[channel];
+        fd = open(in_path, O_RDONLY);
+        continue;
+      }
+    } else {
+      results[channel]->n_input_chunks++;
+      chunk_channel[chunk_idx] = channel;
+    }
+
+    if (bytes_in_window > 0 && window[0] != '@') {
+      // TODO: Perhaps count the lines here to report a more useful error message.
+      cerr << chrono_time() << ":  "
+           << "Error:  File does not conform to FASTQ specification at position " << total_chars << ": " << in_path << endl;
+      exit(EXIT_FAILURE);
+    }
+
+    total_chars += bytes_read;
+
+    char *last_read_addr = last_read(window, bytes_in_window);
+
+    if (bytes_in_window > 0 && last_read_addr == NULL) {
+      // TODO: Count lines here to report a more useful error message.
+      cerr << chrono_time() << ":  "
+           << "Error:  File does not conform to FASTQ specification in the window between characters "
+           << (total_chars - bytes_read) + (last_read_addr - window) << " and " << total_chars << ": " << in_path << endl;
+      exit(EXIT_FAILURE);
+    }
+
+    const auto bytes_until_last_read = last_read_addr - window;
+    const auto bytes_in_chunk = (bytes_in_window < segment_size) ? bytes_in_window : bytes_until_last_read;
+    actual_size[chunk_idx] = bytes_in_chunk;
+    bytes_leftover = bytes_in_window - bytes_in_chunk;
+    const auto leftover_addr = window + bytes_in_chunk;
+
+    // this function will output result for an input file
+    auto write_output_func = [&](const int result_idx) {
+      auto &r = *results[result_idx];
+      r.write_output();
+      {
+        unique_lock<mutex> lk(mtx); // this acquires the mutex mtx
+        --n_running_threads;
+        lk.unlock(); // this explicit unlock is a perf optimization of some sort, see "condition_variable" docs for C++11
+        cv.notify_one();
+      }
+    };
+
+    // this function will run in each query thread
+    auto run_queries = [&](int my_chunk_idx) {
+      uint64_t n_reads;
+      {
+        vector<uint64_t> kmt;
+        n_reads = kmer_lookup_chunk(&kmt, lmer_index, mmer_bloom, kmers_index, snps, buffer_addr + segment_size * my_chunk_idx,
+                                    actual_size[my_chunk_idx], M2, M3, in_path, s_start);
```

Comment:
`in_path` here needs to be replaced with `results[chunk_channel[my_chunk_idx]]->in_path`

this only matters when we have multiple input files in progress on different threads and one of them contains malformatted fastq --- the fix makes sure the error message cites the correct file

i've made the correction in a subsequent PR